### PR TITLE
test: Add unit tests to CI

### PR DESCRIPTION
**Pull request by 8me**
Started to add some unit tests (including an updated CI script) and coverage check for the code.

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Require scikit-hep-testdata>=0.3.1 for LHE file
* Add tests run with pytest
* Add coverage summary and checking with Codecov
   - Add associated config files: codecov.yml, pytest.ini, .coveragerc
* Add MANIFEST.in check to publishing workflow
```

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -27,6 +27,24 @@ jobs:
         python -m pip install --upgrade pip setuptools wheel
         python -m pip install -q --no-cache-dir -e .[complete]
         python -m pip list
+    - name: Unit Tests
+      run: |
+        python -m pip install pytest
```
Comment (lukasheinrich) [inline .github/workflows/ci.yml:32]: could we add this to a `[test]` extra? We do something similar in pyhf

https://github.com/scikit-hep/pyhf/blob/master/setup.py#L25

File: tests/test_lhe_reader.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+__email__ = "jschumann@km3net.de"
+__status__ = "Development"
+
+import unittest
+import pylhe
+from os.path import dirname, join
+
+TEST_DATA_DIR = join(dirname(__file__), 'test_data')
+
+class TestLHEFile(unittest.TestCase):
```
Comment (lukasheinrich) [inline tests/test_lhe_reader.py:19]: do we want to go with `unittest` or  just `pytest`  auto-discovery (I'm somewhat partial to the latter?)

File: tests/test_lhe_reader.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Filename: test_lhe_reader.py
+
+__author__ = "Johannes Schumann"
```
Comment (lukasheinrich) [inline tests/test_lhe_reader.py:5]: I think we might need to remove these since the license etc is handled by LICENSE and it'd be difficult to track the authors this way (git blame should wook)

**Comment by lukasheinrich** [comment]
Hi @8me thanks a lot for this nice PR! This is sorely needed. I added a few comments. Also pinging @matthewfeickert in case he wants to take a look

**Comment by lukasheinrich** [comment]
as for the test LHE file, maybe here we can use https://github.com/scikit-hep/scikit-hep-testdata. Tagging @jpivarski to see what the process wuold be

File: tests/test_lhe_reader.py (hunk 3 above)
Comment (8me) [reply tests/test_lhe_reader.py:5]: Yeah no problem!

**Comment by jpivarski** [comment]
The files should be less than 10 MB (preferably less than 1 MB) each; there's a [60 MB total limit on PyPI packages](https://github.com/scikit-hep/scikit-hep-testdata/issues/30#issuecomment-628023098) (which may be negotiable).

Given that the files are small and test features, rather than bulk size, you can add them as a PR to the [src/skhep_testdata/data directory](https://github.com/scikit-hep/scikit-hep-testdata/tree/master/src/skhep_testdata/data). Once accepted and pushed to a new version, they're accessible in your tests like this

https://github.com/scikit-hep/uproot4/blob/1c4203593ebca427defb200fde81c6ac42ff3d20/tests/test_0008-start-interpretation.py#L35

If you're using Python 3.5, you need to install scikit-hep-testdata with pip, rather than conda (see scikit-hep/scikit-hep-testdata#31).

File: .github/workflows/ci.yml (hunk 1 above)
Comment (matthewfeickert) [reply .github/workflows/ci.yml:32]: I would strongly recommend doing exactly this (in a separate PR at these should be atomic).

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -27,6 +27,24 @@ jobs:
...
+        py.test --junitxml=./reports/junit.xml -o junit_suite_name=pylhe $GITHUB_WORKSPACE
```
Comment (matthewfeickert) [inline .github/workflows/ci.yml:33]:
```suggestion
        python -m pytest --junitxml=./reports/junit.xml -o junit_suite_name=pylhe $GITHUB_WORKSPACE
```

`pytest` should always be run with `python -m pytest` as we have observed multiple times over the years that `py.test` does not always give the expected runtime behaviour.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -27,6 +27,24 @@ jobs:
...
+    - name: Coverage
+      if: matrix.python-version == 3.7 && matrix.os == 'ubuntu-latest'
+      run: |
+        python -m pip install pytest-cov
+        py.test --cov $GITHUB_WORKSPACE --cov-report term-missing --cov-report xml:reports/coverage.xml --cov-report html:reports/coverage
```
Comment (matthewfeickert) [inline .github/workflows/ci.yml:38]:
```suggestion
        python -m pytest --cov $GITHUB_WORKSPACE --cov-report term-missing --cov-report xml:reports/coverage.xml --cov-report html:reports/coverage
```

File: tests/test_lhe_reader.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+__copyright__ = "Copyright 2020, Johannes Schumann"
+__credits__ = []
+__license__ = "MIT"
+__maintainer__ = "Johannes Schumann"
+__email__ = "jschumann@km3net.de"
+__status__ = "Development"
```
Comment (matthewfeickert) [inline tests/test_lhe_reader.py:11]: This should all get removed
```suggestion
```

File: tests/test_lhe_reader.py (hunk 2 above)
Comment (matthewfeickert) [reply tests/test_lhe_reader.py:19]: I would highly recommend this is all done in `pytest`.

Comment (8me) [reply tests/test_lhe_reader.py]: See [9d9998d](https://github.com/scikit-hep/pylhe/pull/29/commits/9d9998d2e40e754bf1333f719e1444de040b2249)

File: .github/workflows/ci.yml (hunk 1 above)
Comment (matthewfeickert) [reply .github/workflows/ci.yml]: This is done in PR #32

File: pytest.ini [hunk 7]
```diff
@@ -0,0 +1,5 @@
+[pytest]
+docstyle_convention = numpy
+junit_family=xunit2
+python_files = test_*.py
+python_functions = test_*
```
Comment (matthewfeickert) [inline pytest.ini:5]: Why are these needed?

File: tests/test_lhe_reader.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python
+# coding=utf-8
+
```
Comment (matthewfeickert) [inline tests/test_lhe_reader.py:3]: These are never going to be run in a standalone manner, so these can get removed.

File: pytest.ini (hunk 7 above)
Comment (8me) [reply pytest.ini:5]: I thought those would be nice to have a fixed definition what is used for testing by `pytest`.

File: tests/test_lhe_reader.py (hunk 8 above)
Comment (8me) [reply tests/test_lhe_reader.py]: Ok, normally I add them just to be sure ;-) ... I have removed them in [10b08ee](https://github.com/scikit-hep/pylhe/pull/29/commits/10b08ee2b5668aafbf1fbc79db22841478b09744)

File: pytest.ini (hunk 7 above)
Comment (matthewfeickert) [reply pytest.ini:5]: Isn't this automatically done by pytest? If you run `python -m pytest` in a top level directory of a repo it doesn't attempt to run every Python file.

Comment (8me) [reply pytest.ini:5]: In my standard setup for unit testing in my projects I use `unittest` as it was in the beginning, so I was not sure about the conventions in this configuration. I added them just to be sure ... I think you're familiar with it, should I just drop them?

**Comment by 8me** [comment]
You're welcome! I'm glad for that `pylhe` is available, so trying to improve it is the least I can do ;-)
The black style check error was caused by some sloppy configuration in my setup ... so with 
[8cc7604](https://github.com/scikit-hep/pylhe/pull/29/commits/8cc76048771e4512a1c2648bd74e90ba5f0d0138) this should be fixed.
Yeah I think also that this error is caused by the missing patch release, but no reason to push some work-around for something that fixes itself in a little while.

File: pytest.ini (hunk 7 above)
Comment (lukasheinrich) [reply pytest.ini:5]: I think this should all be covdred by the default

Comment (8me) [reply pytest.ini]: Ok, I removed it in [8ece4bb](https://github.com/scikit-hep/pylhe/pull/29/commits/8ece4bbf7dedba99c0bb4376a3bed19812d957f1)!

**Comment by 8me** [comment]
Regarding the polite question about the new patch release of `scikit-hep-testdata` ... ping @jpivarski, @lukasheinrich :-D

**Comment by jpivarski** [comment]
@benkrikler @eduardo-rodrigues Who is the maintainer of scikit-hep-testdata (as in, who deploys releases)?

**Comment by eduardo-rodrigues** [comment]
@benkrikler is the maintainer of `scikit-hep-testdata`  but anyone in the org should be able to make a release and help out. I'm happy to make the release and deploy if you need it quickly, which I believe you do in order to make a major release of this package. Right?

**Comment by jpivarski** [comment]
> @benkrikler is the maintainer of `scikit-hep-testdata` but anyone in the org should be able to make a release and help out. I'm happy to make the release and deploy if you need it quickly, which I believe you do in order to make a major release of this package. Right?

Even though we all have permissions, I think we should define a canonical maintainer for each repo, just for coordination. Without that, two people might try to pull the lever at once, zero people might pull the lever (each thinking someone else is going to do it), or one person pulling the lever might upset someone else's plans (someone might be working to get a bug-fix into a release before it affects users downstream). This maintainer isn't on the hook to fix all bugs, just to be the single pair of eyes to prevent collisions.

**Comment by eduardo-rodrigues** [comment]
> Even though we all have permissions, I think we should define a canonical maintainer for each repo, just for coordination. Without that, two people might try to pull the lever at once, zero people might pull the lever (each thinking someone else is going to do it), or one person pulling the lever might upset someone else's plans (someone might be working to get a bug-fix into a release before it affects users downstream). This maintainer isn't on the hook to fix all bugs, just to be the single pair of eyes to prevent collisions.

@jpivarski, I don't think I said anything in disagreement with you, no? I did say *to help out* and that Ben is the maintainer. But if you take it that way I'm also happy to do nothing. In any case Ben seems out-of-reach these days ...

**Comment by jpivarski** [comment]
> @jpivarski, I don't think I said anything in disagreement with you, no? I did say *to help out* and that Ben is the maintainer. But if you take it that way I'm also happy to do nothing. In any case Ben seems out-of-reach these days ...

I thought you were saying that any of us could/should make the release. I'm saying that this could be too chaotic. If we're going to help out, at least let's make sure to have some sort of sign-up sheet so that we know who to call.

**Comment by eduardo-rodrigues** [comment]
That's fine by me. Take the example of Particle and DecayLanguage. Both Henry and I make commits, review each other and make releases according to availability of each of us. Work well.

**Comment by 8me** [comment]
How what is the current status, I think the testfile should be included in `scikit-hep-testdata` since v0.3.1?!

**Comment by matthewfeickert** [comment]
> How what is the current status, I think the testfile should be included in `scikit-hep-testdata` since v0.3.1?!

Ah great. Can you rebase against `master` and then in `setup.py` require

```python
"scikit-hep-testdata>0.3.0"
```
? We can go from there again. Sorry I think we all got sidetracked here.

**Comment by matthewfeickert** [comment]
@8me It seems that something went wrong here. Can you please make sure that you have fetched against the upstream `master` and then rebase? Let us know if you have questions on how to do this.

**Comment by 8me** [comment]
Yeah I forgot to sync my fork ... sorry for that and the confusion, I'm already over it to fix that! I will comment again when I'm finished!

**Comment by 8me** [comment]
After I mixed up the hash for the revoke I set it back manually by doing a checkout. I hope the rebase is now correct as the rebase throw some errors because of the `ci.yml`. I added the unittest and coverage step again, but to be honest I was not sure what was the final decision regarding this ... hope I remembered it correctly.

**Comment by matthewfeickert** [comment]
Something has gone very wrong with the rebase. The GitHub UI now thinks that many files are being created which should not have been touched by this PR at all &mdash; this make review hard. It is also unclear what has happened given that there are now 186 commits that are on your branch

```
$ git rev-list --count 8m3/testing origin/master
186
```

My suggestion is that you make a new branch and use `git reset` to reset the state of that branch to somewhere that makes sense to you. From there, I'l make you understand what all is on the branch and then you can a hard of reset this PR branch to that one. Alternatively, if you know the files that you want, just make a new branch and then use `git checkout` to get the state of those files into the new branch, and then force the state of that branch for this PR.

**Comment by 8me** [comment]
Yeah think there is no other option than your suggestion how to handle it ... I thought I went through the rebase by the book and I thought I got it first. I think I'll open a new branch for me in order not to make it even worse. I'd say give me some days for me to get my lesson about 'how to rescue my PR after I totally blew it up' ;-) ... I will call for help if get totally stuck. Sorry again for annoying you with me being a little bit toooo careless with `git rebase`!

**Comment by matthewfeickert** [comment]
> Sorry again for annoying you

You're not annoying us! You're donating you time to this project and we appreciate it. I'm sorry this is happening to you at the moment, but I think everyone has had a few `rebases` go sidways in their life. :)

**Comment by 8me** [comment]
Ok I might have something now, maybe you can have a look at [8me/pylhe/testing-backup](https://github.com/8me/pylhe/tree/testing-backup) what I did:
1. Go back to an ancient commit in the `testing` branch when the world was still ok and created a new branch: `it checkout 592b4c58bdf3a2655964e5249be3780cb5ec004a` & `git branch testing-backup`
2. Re-run the rebase with theirs option, to trust master (which is mostly needed because `ci.yml`): `git rebase master -X theirs`
3. Updated the `ci.yml` and the `setup.py` manually as it was requested
The check via `git log --oneline testing-backup ^origin/master | wc -l` gives me now 28. In principle this is what I already tried to do yesterday, but I really don't know which step blew it up at the first try.

**Comment by matthewfeickert** [comment]
> Ok I might have something now

Yup! When I look at `https://github.com/scikit-hep/pylhe/compare/master...8me:testing-backup` that all seems perfect! :rocket:

So I guess what I would say now is just do a hard reset of your `testing` branch to your `testing-backup` branch's state. So something like

```
git checkout origin/testing -b restore-if-all-fails # make a backup of the current state of this PR
git checkout testing
git reset --hard origin/testing-backup # change the state of the branch to be testing-backup
git push --force-with-lease
```

If that all goes well then we can just carry on here for any final touch ups and then you can delete the `restore-if-all-fails` at your leisure. :)

File: MANIFEST.in [hunk 9]
```diff
@@ -1,5 +1,7 @@
 graft src
 
 include LICENSE
+exclude pytest.ini
 
+global-exclude test_*.py
```
Comment (matthewfeickert) [inline MANIFEST.in:6]: I think that these aren't actually needed. `graft src` above is only going to add the contents of `src/` to the MANIFEST and everything else will be ignored by default.  Can you confirm? If the `check-manifest` is erroring then things can be added to https://github.com/scikit-hep/pylhe/blob/087f6abd0ed2c5e3c201938fbcc155650d5be1e4/pyproject.toml#L23-L30
```suggestion
```

Comment (8me) [reply MANIFEST.in:6]: Looks good: `check-manifest` quotes the files as 'missing from sdist'!

**Comment by matthewfeickert** [comment]
@8me To get Codecov working requires some tweaking so I just am pushing here to verify somethings &mdash; everything is looking good now on your side. @lukasheinrich I think this can get reviewed again and then once we have Codedov happy I think we can merge. :+1:

**Comment by matthewfeickert** [comment]
@lukasheinrich I left a suggested commit message for the squash commit in the PR description body.

**Comment by lukasheinrich** [comment]
Thanks @8me for this nice PR. Finally we have testing in pylhe :)
